File: Stack/503.Next-Greater-Element-II/503.Next-Greater-Element-II.py
class Solution:
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        stack=[]
        res = [-1]*len(nums)
        for i in list(range(len(nums))) * 2:
            while (stack and nums[stack[-1]]<nums[i]):
                res[stack.pop()]=nums[i]                
            stack.append(i)
        return res

File: Stack/503.Next-Greater-Element-II/503.Next-Greater-Element-II.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) 
    {
        int N = nums.size();
        for (int i=0; i<N; i++)
            nums.push_back(nums[i]);
        
        stack<int>Stack;        
        vector<int>results(N,-1);
        
        for (int i=0; i<nums.size(); i++)
        {
            while (!Stack.empty() && nums[Stack.top()]<nums[i])
            {
                if (Stack.top() < N) 
                    results[Stack.top()] = nums[i];
                Stack.pop();
            }
            Stack.push(i);            
        }
        results.resize(N);
        return results;
    }
};
